cheetah/accelerator/space_charge_kick.py

```diff
@@ -0,0 +1,610 @@
+        )
+        potential_ft = charge_density_ft * integrated_green_function_ft
+        potential = (1 / (4 * torch.pi * epsilon_0)) * torch.fft.ifftn(
+            potential_ft, dim=[1, 2, 3]
+        ).real
+
+        # Return the physical potential
+        return potential[
+            :,
+            : charge_density.shape[1] // 2,
+            : charge_density.shape[2] // 2,
+            : charge_density.shape[3] // 2,
+        ]
+
+    def _E_plus_vB_field(
+        self, beam: ParticleBeam, cell_size: torch.Tensor, grid_dimensions: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Computes the force field from the potential and the particle positions and
+        speeds, as in https://doi.org/10.1063/1.2837054.
+        """
+        inv_cell_size = 1 / cell_size
+        gamma = self._gammaref(beam)
+        igamma2 = 1 / gamma**2 if gamma != 0 else torch.tensor(0.0)
+        potential = self._solve_poisson_equation(beam, cell_size, grid_dimensions)
+
+        grad_x = torch.zeros_like(potential)
+        grad_y = torch.zeros_like(potential)
+        grad_s = torch.zeros_like(potential)
+
+        # Compute the gradients of the potential, using central differences, with 0
+        # boundary conditions.
+        grad_x[:, 1:-1, :, :] = (potential[:, 2:, :, :] - potential[:, :-2, :, :]) * (
+            0.5 * inv_cell_size[:, 0, None, None, None]
+        )
+        grad_y[:, :, 1:-1, :] = (potential[:, :, 2:, :] - potential[:, :, :-2, :]) * (
+            0.5 * inv_cell_size[:, 1, None, None, None]
+        )
+        grad_s[:, :, :, 1:-1] = (potential[:, :, :, 2:] - potential[:, :, :, :-2]) * (
+            0.5 * inv_cell_size[:, 2, None, None, None]
+        )
+
+        # Scale the gradients with lorentz factor
+        grad_x = -igamma2[:, None, None, None] * grad_x
+        grad_y = -igamma2[:, None, None, None] * grad_y
+        grad_s = -igamma2[:, None, None, None] * grad_s
+
+        return grad_x, grad_y, grad_s
+
+    def _cheetah_to_moments(self, beam: ParticleBeam) -> torch.Tensor:
+        """
+        Converts the Cheetah particle beam parameters to the moments in SI units used
+        in the space charge solver.
+        """
+        moments = beam.particles
+        gammaref = self._gammaref(beam)
+        betaref = self._betaref(beam)
+        p0 = gammaref * betaref * electron_mass * c
+        gamma = gammaref[:, None] * (
+            torch.ones(moments.shape[:-1]) + beam.particles[:, :, 5] * betaref[:, None]
+        )
+        beta = torch.sqrt(1 - 1 / gamma**2)
+        p = gamma * electron_mass * beta * c
+        moments[:, :, 1] = p0[:, None] * moments[:, :, 1]
+        moments[:, :, 3] = p0[:, None] * moments[:, :, 3]
+        moments[:, :, 4] = -betaref[:, None] * moments[:, :, 4]
+        moments[:, :, 5] = torch.sqrt(
+            p**2 - moments[:, :, 1] ** 2 - moments[:, :, 3] ** 2
+        )
+
+    def _moments_to_cheetah(self, beam: ParticleBeam) -> torch.Tensor:
```

Comment:
This function currently doesn't return anything ... it seems to do things inplace 🤔